src/3devoUpdate/DeviceList.cs

```diff
@@ -0,0 +1,156 @@
+      string DeviceSearchQuery = "SELECT * FROM Win32_USBControllerDevice";
+      ManagementObjectSearcher USBControllerDeviceCollection = new ManagementObjectSearcher(DeviceSearchQuery);
+
+      if( USBControllerDeviceCollection != null ) {
+
+        foreach( ManagementObject queryObj in USBControllerDeviceCollection.Get() ) {
+          // Get the DeviceID of the device entity
+          string Dependent = (queryObj["Dependent"] as string).Split(new Char[] { '=' })[1];
+
+          // Filter out USB devices without VID and PID
+          Match match = Regex.Match(Dependent, "VID_[0-9|A-F]{4}&PID_[0-9|A-F]{4}");
+          if( match.Success ) {
+            UInt16 VendorID = Convert.ToUInt16(match.Value.Substring(4, 4), 16); // Vendor ID
+            if( FilterVid != UInt16.MinValue && FilterVid != VendorID ) continue;
+
+            UInt16 ProductID = Convert.ToUInt16(match.Value.Substring(13, 4), 16); // Product Number
+            if( FilterPid != UInt16.MinValue && FilterPid != ProductID ) continue;
+
+            // Check to see if our interface number is available for uploads
+            Match matchInterface = Regex.Match(Dependent, "VID_[0-9|A-F]{4}&PID_[0-9|A-F]{4}&MI_[0-9|A-F]");
+            if( matchInterface.Success ) {
+              continue; // No need for the interface sub-nodes
+            } else {
+              // The match fails if the interface is not available, meaning that this is the "parent" node.
+              // When this is the case, the serial number will be available at the end of the string.
+              UInt64 SerialNum = Convert.ToUInt64(Dependent.Substring(25, 12), 16);
+
+              string PnPEntrySearchQuery = "SELECT * FROM Win32_PnPEntity WHERE DeviceID=" + Dependent;
+              ManagementObjectCollection PnPEntityCollection = new ManagementObjectSearcher(PnPEntrySearchQuery).Get();
+
+              if( PnPEntityCollection != null ) {
+                foreach( ManagementObject Entity in PnPEntityCollection ) {
+                  USBDeviceList Element = new USBDeviceList(
+                      "" as string, // Device ID not available right now
+                      Entity["PNPDeviceID"] as string, // PnP Device ID
+                      Entity["Description"] as string // Device Description
+                    );
+
+                  Guid ClassGuid = new Guid(Entity["ClassGuid"] as String); // Device installation class GUID
+                  if( GUID_USB_CLASS_USB_DEVICE != Guid.Empty && ClassGuid == GUID_USB_CLASS_USB_DEVICE ) {
+                    Element.MachineName = USBDeviceList.MachineType.StBootloader; // Uses a generic USB device ClassGUID
+                  } else if( GUID_USB_CLASS_BUS_DEVICES != Guid.Empty && ClassGuid == GUID_USB_CLASS_BUS_DEVICES ) {
+                    Element.MachineName = USBDeviceList.MachineType.AiridDryer; // Makes use of WinUsb driver ClassGuid

```

Comment:
It seems weird to semi-hardcode the `MachineName` here (and above for serial devices). I believe the device name should really just be passed as an argument to `GetDfuDevices` or `GetSerialDevices`, since it just depends on the VID/PID. Alternatively, it could be left out and the caller could just look at the VID/PID directly maybe.